clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/Types.hs

```diff
@@ -0,0 +1,161 @@
+-- | Contains the types used in the implementation of the GMII to SGMII bridge wrapper.
+-- LogiCORE IP Ethernet 1000BASE-X PCS/PMA or SGMII
+
+{-# LANGUAGE LambdaCase #-}
+{-# OPTIONS_GHC -fconstraint-solver-iterations=10 #-}
+{-# LANGUAGE RecordWildCards #-}
+module Clash.Cores.Xilinx.Ethernet.Gmii.Types where
+
+import Clash.Prelude
+
+data Lvds = Lvds{lvdsP :: Bit, lvdsN :: Bit}
+  deriving (Generic, NFDataX, BitPack)
+
+-- TODO: Incorporate in 'Clash.Prelude'?
+-- | Gets time in 'Picoseconds' from time in 'Nanoseconds'.
+type Nanoseconds  (ns :: Nat) = Picoseconds  (1000 * ns)
+-- | Gets time in 'Picoseconds' from time in 'Picoseconds', essentially 'id'.
+type Picoseconds  (ps :: Nat) = ps
+-- | A generic GMII interface
+
+-- TODO: Move to re-usable module?
+-- | Represents a GMII interface.
+data Gmii = Gmii
+ { gmiiData :: BitVector 8
+ , gmiiValid :: Bit
+ , gmiiError :: Bit
+ } deriving (Generic, BitPack, NFDataX)
+
+data DuplexMode = HalfDuplex | FullDuplex
+  deriving (Generic, NFDataX, BitPack, Eq)
+
+-- | Link speeds in Mega bits per second.
+data LinkSpeed = Speed10 | Speed100 | Speed1000
+  deriving (Generic, NFDataX, BitPack)
+
+-- | Whether the link partner supports pause frames.
+data Pause = NoPause | SymmetricPause | AsymmetricPause | SymmetricAsymmetricPause
+  deriving (Generic, NFDataX, BitPack)
+
+type AutoNegConfigVector = BitVector 16
+-- | A subset of the an_adv_config_vector that is applicable for the GMII to SGMII bridge
+data AutoNegConfig =
+  AutoNegConfig
+  { cLinkSpeed :: LinkSpeed
+  , cDuplexMode :: DuplexMode
+  } deriving (Generic, NFDataX, BitPack)
+
+-- | Converts the AutoNegConfig Records to the an_adv_config_vector format.
+toAutoNegConfigVector :: AutoNegConfig -> AutoNegConfigVector
+toAutoNegConfigVector AutoNegConfig{..} = bv4
+  where
+    (linkSpeed0, linkSpeed1) = case cLinkSpeed of
+      Speed10 -> (0, 0)
+      Speed100 -> (0, 1)
+      Speed1000 -> (1, 0)
+
+    bv0 = 0x4001 -- How the bridge is configured in MAC mode.
+
+    -- Apply the configuration.
+    bv1 = replaceBit (12 :: Int) (bitCoerce $ cDuplexMode == FullDuplex) bv0
+    bv2 = replaceBit (10 :: Int) linkSpeed0 bv1
+    bv3 = replaceBit (11 :: Int) linkSpeed1 bv2
+
+    -- Set link status
+    bv4 = setBit bv3 (15:: Int)
+
+
+-- | 10 Mbit and half duplex are the default values since they are the most likely
+-- to be supported by the link partner.
+instance Default AutoNegConfig where
+  def = AutoNegConfig {..}
+   where
+    cLinkSpeed = Speed10
+    cDuplexMode = HalfDuplex
+
+type ConfigVector = BitVector 5
+-- | A record representation of the configuration_vector.
+data Config = Config
+  { cUniDirectional :: Bool
+  , cLoopback :: Bool
+  , cPowerDown :: Bool
+  , cIsolateGmii :: Bool
+  , cAutoNegEnable :: Bool
+  } deriving (Generic, NFDataX, BitPack)
+
+-- | Transform the Config record to a configuration_vector.
+toConfigVector :: Config -> ConfigVector
+toConfigVector Config{..} = pack
```

Comment:
Like the other `pack` suggestion, personally I'd prefer to not implicitly rely on the order of constructors, or in this case, field definitions.

Reply:
To that I have two counterarguments:
1. I don't think it's actually implicit. I mean, our documentation doesn't do enough to call it out, but ADTs always generate `BitPack` instances that follow [`ConstructorType Binary`](https://hackage.haskell.org/package/clash-prelude-1.8.1/docs/Clash-Annotations-BitRepresentation-Deriving.html#t:ConstructorType) and [`FieldsType OverlapL`](https://hackage.haskell.org/package/clash-prelude-1.8.1/docs/Clash-Annotations-BitRepresentation-Deriving.html#t:FieldsType).
2. Making your ADTs match some predetermined layout is good practice because it saves LUTs. In the `Config` case, it renders to wires, but for instance in the case of `LinkSpeed` it actually leads to LUTs that are completely superfluous. And the alternative of explicitly generating a `BitRepresentation` that matches what it automatically would be anyway is just unnecessary complexity, both to the compiler as well as to the programmer.

I'm of the belief that labeling this "implicitly relying on order of constructors" is a misconception and an impediment to good design. But we can widen the discussion up and ask more of our colleagues what they think.

Reply:
> ... but ADTs always generate BitPack instances that follow [ConstructorType Binary](https://hackage.haskell.org/package/clash-prelude-1.8.1/docs/Clash-Annotations-BitRepresentation-Deriving.html#t:ConstructorType) and [FieldsType OverlapL](https://hackage.haskell.org/package/clash-prelude-1.8.1/docs/Clash-Annotations-BitRepresentation-Deriving.html#t:FieldsType).

So the order of constructors determines the meaning as interpreted by the IP, which is what I wanted to prevent.

> Making your ADTs match some predetermined layout is good practice because it saves LUTs.
I do agree with you, perhaps we can do it that way instead and create some small Hunit assertions that verify the correctness?

Reply:
> So the order of constructors determines the meaning as interpreted by the IP, which is what I wanted to prevent.

I'm not sure I follow, but that sounds in reverse. The interpretation by the IP determines the order you specify the constructors. The IP is what's fixed, it's the determinant.

> perhaps we can do it that way instead and create some small Hunit assertions that verify the correctness?

That'd be nice, yes.